test_conformance/api/test_api_consistency.cpp

```diff
@@ -473,14 +437,689 @@ int test_consistency_pipes(cl_device_id deviceID, cl_context context,
+
+int test_consistency_device_and_host_timer(cl_device_id deviceID,
+                                           cl_context context,
+                                           cl_command_queue queue,
+                                           int num_elements)
+{
+    // clGetPlatformInfo, passing CL_PLATFORM_HOST_TIMER_RESOLUTION
+    // May return 0, indicating that platform does not support Device and Host
+    // Timer Synchronization.
+    cl_int error;
+
+    cl_platform_id platform = NULL;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_PLATFORM, sizeof(platform),
+                            &platform, NULL);
+    test_error(error, "Unable to query CL_DEVICE_PLATFORM");
+
+    cl_ulong hostTimerResolution = 0;
+    error = clGetPlatformInfo(platform, CL_PLATFORM_HOST_TIMER_RESOLUTION,
+                              sizeof(hostTimerResolution), &hostTimerResolution,
+                              NULL);
+    test_error(error, "Unable to query CL_PLATFORM_HOST_TIMER_RESOLUTION");
+
+    if (hostTimerResolution == 0)
+    {
+        // clGetDeviceAndHostTimer, clGetHostTimer
+        // Returns CL_INVALID_OPERATION if the platform associated with device
+        // does not support Device and Host Timer Synchronization.
+
+        cl_ulong dt = 0;
+        cl_ulong ht = 0;
+
+        error = clGetDeviceAndHostTimer(deviceID, &dt, &ht);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_PLATFORM_HOST_TIMER_RESOLUTION returned 0 but "
+            "clGetDeviceAndHostTimer did not return CL_INVALID_OPERATION");
+
+        error = clGetHostTimer(deviceID, &ht);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_PLATFORM_HOST_TIMER_RESOLUTION returned 0 but "
+            "clGetHostTimer did not return CL_INVALID_OPERATION");
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_il_programs(cl_device_id deviceID, cl_context context,
+                                 cl_command_queue queue, int num_elements)
+{
+    // clGetDeviceInfo, passing CL_DEVICE_IL_VERSION or
+    // CL_DEVICE_ILS_WITH_VERSION
+    // May return an empty string and empty array, indicating that device does
+    // not support Intermediate Language Programs.
+    cl_int error;
+
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+
+    // Even if the device does not support Intermediate Language Programs the
+    // size of the string query should not be zero.
+    size_t sz = 0;
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_IL_VERSION, 0, NULL, &sz);
+    test_error(error, "Unable to query CL_DEVICE_IL_VERSION");
+    test_assert_error(sz != 0,
+                      "CL_DEVICE_IL_VERSION should return a non-zero size");
+
+    std::string ilVersion = get_device_il_version_string(deviceID);
+
+    error = clGetDeviceInfo(deviceID, CL_DEVICE_ILS_WITH_VERSION, 0, NULL, &sz);
+    test_error(error, "Unable to query CL_DEVICE_ILS_WITH_VERSION");
+
+    if (ilVersion == "" || sz == 0)
```

Comment:
`CL_DEVICE_ILS_WITH_VERSION` may return an empty array, not an empty string.

I think this should be updated to follow a similar path to the extended versioning test, making sure that a returned `cl_name_version_khr` is empty.

Reply:
Actually, I think this highlights a problem in extended versioning, not this consistency test.

I'll go and check and then update here.

Reply:
Cool, I do think this code is correct.  There are two cases to consider:

1. What does `CL_DEVICE_IL_VERSION` return?  This is the single string query added in OpenCL 2.1.  This string is either empty (`""`, indicating that no ILs are supported) or non-empty (ILs supported), but it should not return size equal to zero.

2. What does `CL_DEVICE_ILS_WITH_VERSION` return?  This is the new OpenCL 3.0 array query.  This query may return size equal to zero (no array elements, indicating that no ILs are supported) or a non-zero size (some array elements, indicating that ILs are supported).

The check for `ilVersion == ""` is checking for the first case, and the check for `sz == 0` is checking for the second case.

Reply:
Perhaps this is where the confusion is coming from (empty string handling).

What exactly do we say the empty string is? In C its just `\0` which has a length of 0. That would mean that `sz` *can* be 0, but that is asserted against above.

Reply:
(For reference, we currently return a size of 1 and have it simply contain `\0`)

Reply:
I believe the issue is this:

Any device that does not support `CL_DEVICE_IL_VERSION` will fill out `1` for `param_value_size_ret`. Subsequently, `param_value` will be 1 byte and just contain `\0`.

`get_device_il_version_string` calls `get_device_info_string` which constructs an `std::string` using this data, resulting in a string whose length is not 0. Therefore, the constructed string is not equivalent to an `std::string("")`, so `ilVersion == ""` evaluates to `false`.

If my understanding is correct and I'm not missing something obvious (please correct me if I'm wrong), then I think this can be avoided in one of two ways.

1. `get_device_info_string` can be special cased for empty strings
2. The comparison in this test can be changed from `ilVersion == ""` to `strcmp(ilVersion.c_str(), "") == 0`

My preference would be to go with 1. (possibly as a separate PR), as the empty string case may crop up in other tests in the future and is not specific to this test.

Reply:
Ah, I see.  Yes, we are constructing our `std::string` with:

```c
    std::vector<char> info(size);

    if ((err = clGetDeviceInfo(device, param_name, size, info.data(), NULL))
        != CL_SUCCESS)
    {
        throw std::runtime_error("clGetDeviceInfo failed\n");
    }

    return std::string(info.begin(), info.end());
```

I think we could change this to either `std::string(info.begin(), --info.end());` or `std::string(info.data(), size - 1)`, and get the behavior we are looking for.  Any preference?  I'm leaning towards the latter.

Reply:
No strong preference, so the latter looks good to me. Thanks.

Would you like me to do that as part of a new PR, or just do it with this PR?

Reply:
I'll take care of it.  Thanks for the quick reply!

Reply:
Thanks @bashbaug, looks good.